### PR TITLE
Support to paths in the Private URL of the APIs

### PR DESCRIPTION
**Pull request by guicassolato**
**What this PR does / why we need it**:
It makes APIs' Private URLs to support a "path" part, changing the accepted format from `http(s)://host(:port)` to `http(s)://host(:port)(/path)`.

**Which issue(s) this PR fixes** 
Closes https://github.com/3scale/porta/issues/2

**Special notes for your reviewer**:
Possible implications to https://github.com/3scale/apicast should also be considered.

### CODE REVIEW

File: app/validators/uri_validator.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    false
+  end
+
+  def valid_scheme?(scheme)
+    accepted_scheme = options.fetch(:scheme, DEFAULT_ACCEPTED_SCHEMES)
+    case accepted_scheme
+    when Regexp
+      scheme =~ accepted_scheme
+    else
+      [*accepted_scheme].include? scheme
```
Comment (hallelujah) [inline app/validators/uri_validator.rb]: So it can be an Array or a Regexp?
Why not using only an Array?

File: app/validators/uri_validator.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+class UriValidator < ActiveModel::EachValidator
+  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
+  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
+  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
+
+  def validate_each(record, attribute, value)
+    uri = URI.parse(value)
+    raise URI::InvalidURIError if !valid_scheme?(uri.scheme) || uri.host.blank? || forbidden_part?(uri)
+    true
+  rescue URI::InvalidURIError
```
Comment (hallelujah) [inline app/validators/uri_validator.rb]: Probably its ancestor `URI::Error`

File: app/validators/uri_validator.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class UriValidator < ActiveModel::EachValidator
+  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
+  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
+  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
+
+  def validate_each(record, attribute, value)
+    uri = URI.parse(value)
+    raise URI::InvalidURIError if !valid_scheme?(uri.scheme) || uri.host.blank? || forbidden_part?(uri)
```
Comment (hallelujah) [inline app/validators/uri_validator.rb]: I'd use another class than `InvalidURIError`
Probably not even raising an error here but only setting the error message to the record

File: app/validators/uri_validator.rb (hunk 2 above)
Comment (hallelujah) [reply app/validators/uri_validator.rb]: This is an exception that is separated from the validator, I expect the error message to be different.

1. Validation failed
2. Failed to parse URI (generic unrecoverable error)

Comment (hallelujah) [reply app/validators/uri_validator.rb]: OK reading the `uri/common` and its usage in the ruby stdlib, it looks like using InvalidURIComponent would fit here. Thus raising and rescueing the error inside the same method is really unexpected

File: app/validators/uri_validator.rb (hunk 1 above)
Comment (guicassolato) [reply app/validators/uri_validator.rb]: I think it's a matter of taste more than anything. I was thinking on the usage:

```ruby
validates :field, uri: { scheme: /^[st]?ftp$/ }
```

instead of 

```ruby
validates :field, uri: { scheme: %w[ftp sftp tftp] }
```

But I can change that if you think it's worth. No problem.

File: app/validators/uri_validator.rb (hunk 2 above)
Comment (guicassolato) [reply app/validators/uri_validator.rb]: You are right. It shouldn't raise the exception anyway. I'll fix that.

File: test/unit/validators/uri_validator_test.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class UriValidatorTest < ActiveSupport::TestCase
+  test 'invalid uri' do
+    klass = Class.new(base_record_class) { validates :uri, uri: true }
```
Comment (hallelujah) [inline test/unit/validators/uri_validator_test.rb]: Do you need to subclass the base_record?
Can't you use a fix `Model` class? Note it will be namespaced in `UriValidatorTest::Model`

File: app/validators/uri_validator.rb (hunk 3 above)
Comment (guicassolato) [reply app/validators/uri_validator.rb]: You are right. It shouldn't raise the exception anyway. I'll fix that.

File: test/unit/validators/non_localhost_validator_test.rb [hunk 5]
```diff
@@ -1,10 +1,12 @@
+# frozen_string_literal: true
+
 require 'test_helper'
 
-class ThreeScale::LocalhostValidatorTest < ActiveSupport::TestCase
+class NonLocalhostValidatorTest < ActiveSupport::TestCase
```
Comment (josemigallas) [inline test/unit/validators/non_localhost_validator_test.rb]: Could be worth it adding an edge case to this test, something containing localhost? e.g. `http://mylocalhost/` Could give a false negative, depending on the regex.

Comment (guicassolato) [reply test/unit/validators/non_localhost_validator_test.rb]: I didn't actually wrote that, just renamed it. But I guess you're right. It won't kill me enhancing it a bit since I'm here 😜

Comment (josemigallas) [reply test/unit/validators/non_localhost_validator_test.rb]: I know, I know 😄 .. that's what happens when touching others' stuff hehe

File: app/validators/uri_validator.rb (hunk 2 above)
Comment (guicassolato) [reply app/validators/uri_validator.rb]: `URI.parse` raises `URI::InvalidURIError`.

File: test/unit/validators/non_localhost_validator_test.rb (hunk 5 above)
Comment (guicassolato) [reply test/unit/validators/non_localhost_validator_test.rb]: Done.

File: app/validators/uri_validator.rb (hunk 1 above)
Comment (hallelujah) [reply app/validators/uri_validator.rb]: This list is not changing so much, you just add complexity here I think
